pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java

```diff
@@ -0,0 +1,156 @@
+    _validator = new Validator(SqlStdOperatorTable.instance(), _catalogReader, _typeFactory);
+
+    // optimizer rules
+    _logicalRuleSet = PinotQueryRuleSets.LOGICAL_OPT_RULES;
+
+    // optimizer
+    HepProgramBuilder hepProgramBuilder = new HepProgramBuilder();
+    for (RelOptRule relOptRule : _logicalRuleSet) {
+      hepProgramBuilder.addRuleInstance(relOptRule);
+    }
+    _relOptPlanner = new LogicalPlanner(hepProgramBuilder.build(), Contexts.EMPTY_CONTEXT);
+  }
+
+  public QueryPlan sqlQuery(String sqlQuery) {
+    PlannerContext plannerContext = new PlannerContext();
+    try {
+      SqlNode parsed = this.parse(sqlQuery, plannerContext);
+      SqlNode validated = this.validate(parsed);
+      RelRoot relation = this.toRelation(validated, plannerContext);
+      RelNode optimized = this.optimize(relation, plannerContext);
+      return this.toQuery(optimized, plannerContext);
+    } catch (Exception e) {
+      throw new RuntimeException("Error composing query plan", e);
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // steps
+  // --------------------------------------------------------------------------
+
+  protected SqlNode parse(String query, PlannerContext plannerContext)
+      throws Exception {
+    // 1. invoke CalciteSqlParser to parse out SqlNode;
+    SqlNode compiled = CalciteSqlParser.compile(query, plannerContext);
+    // 2. TODO: add query rewrite logic
+    return compiled;
+  }
+
+  protected SqlNode validate(SqlNode parsed)
+      throws Exception {
+    // 3. validator to validate.
+    SqlNode validated = _validator.validate(parsed);
+    if (null == validated || !validated.getKind().belongsTo(SqlKind.QUERY)) {
+      throw new IllegalArgumentException(
+          String.format("unsupported SQL query, cannot validate out a valid sql from:\n%s", parsed));
+    }
+    return validated;
+  }
+
+  protected RelRoot toRelation(SqlNode parsed, PlannerContext plannerContext) {
+    RexBuilder rexBuilder = new RexBuilder(_typeFactory);
+    RelOptCluster cluster = RelOptCluster.create(_relOptPlanner, rexBuilder);
+    SqlToRelConverter sqlToRelConverter =
+        new SqlToRelConverter(_planner, _validator, _catalogReader, cluster, StandardConvertletTable.INSTANCE,
+            SqlToRelConverter.config());
+    return sqlToRelConverter.convertQuery(parsed, false, true);
+  }
+
+  protected RelNode optimize(RelRoot relRoot, PlannerContext plannerContext) {
+//    RelTraitSet traitSet = relRoot.rel.getCluster().traitSet().replace(EnumerableConvention.INSTANCE).simplify();
+    try {
+      _relOptPlanner.setRoot(relRoot.rel);
+      return _relOptPlanner.findBestExp();
+    } catch (Exception e) {
+      throw new UnsupportedOperationException(
+          "Cannot generate a valid execution plan for the given query: " + RelOptUtil.toString(relRoot.rel), e);
+    }
+  }
+
+  protected QueryPlan toQuery(RelNode relRoot, PlannerContext plannerContext) {
```

Comment:
(nit) not stickler about names but `toQuery()` is probably not intuitive and is not implying the right meaning ?

How about `toPhysicalPlan()` or `toParallelizedPhysicalPlan()`

Reply:
Addressed in latest changes